internal/fwserver/server_planaction.go

```diff
@@ -94,5 +145,24 @@ func (s *Server) PlanAction(ctx context.Context, req *PlanActionRequest, resp *P
 
 		resp.Diagnostics = modifyPlanResp.Diagnostics
 		resp.Deferred = modifyPlanResp.Deferred
+
+		if len(resp.LinkedResources) != len(modifyPlanResp.LinkedResources) {
+			resp.Diagnostics.AddError(
+				"Invalid Linked Resource Plan",
+				"An unexpected error was encountered when planning an action with linked resources. "+
+					fmt.Sprintf(
+						"The number of linked resources produced by the action plan cannot change: %d linked resource(s) were produced in the plan, expected %d\n\n",
+						len(modifyPlanResp.LinkedResources),
+						len(resp.LinkedResources),
+					)+
+					"This is always a problem with the provider and should be reported to the provider developer.",
+			)
+			return
+		}
+
+		for i, newLinkedResource := range modifyPlanResp.LinkedResources {
+			resp.LinkedResources[i].PlannedState = planToState(newLinkedResource.Plan)
```

Comment:
This applies to `InvokeAction` as well, but I'm thinking we might want to consider also:
- Running write-only nullification? (probably both)
- Checking mutability of identities and if they change?
- Running semantic equality logic (probably just in Invoke actually)